src/Nethermind/Nethermind.WebSockets/SocketClientBase.cs

```diff
@@ -0,0 +1,114 @@
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Nethermind.Logging;
+using Nethermind.Serialization.Json;
+
+namespace Nethermind.WebSockets
+{
+    public class SocketClientBase : ISocketsClient
+    {
+        public const int MAX_POOLED_SIZE = 1024 * 1024;
+
+        protected readonly ISocketHandler _handler;
+        protected readonly IJsonSerializer _jsonSerializer;
+
+        public string Id { get; } = Guid.NewGuid().ToString("N");
+        public string ClientName { get; }
+
+        public SocketClientBase(string clientName, ISocketHandler handler, IJsonSerializer jsonSerializer)
+        {
+            ClientName = clientName;
+            _handler = handler;
+            _jsonSerializer = jsonSerializer;
+        }
+
+        public virtual Task ProcessAsync(Memory<byte> data) => Task.CompletedTask;
+
+        public virtual Task SendAsync(SocketsMessage message)
+        {
+            if (message is null)
+            {
+                return Task.CompletedTask;
+            }
+
+            if (message.Client == ClientName || string.IsNullOrWhiteSpace(ClientName) ||
+                string.IsNullOrWhiteSpace(message.Client))
+            {
+                return _handler.SendRawAsync(_jsonSerializer.Serialize(new
+                {
+                    type = message.Type,
+                    client = ClientName,
+                    data = message.Data
+                }));
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public async Task ReceiveAsync()
+        {
+            int currentMessageLength = 0;
+            byte[] buffer = new byte[1024 * 4];
+            byte[] combinedData = Array.Empty<byte>();
+
+            var result = await _handler.GetReceiveResult(buffer);
+            if (result == null)
+            {
+                return;
+            }
+
+            while (!result.Closed)
+            {
+                int newMessageLength = currentMessageLength + result.Read;
+                if (newMessageLength > MAX_POOLED_SIZE)
+                {
+                    throw new InvalidOperationException("Message too long");
+                }
+
+                byte[] newBytes = ArrayPool<byte>.Shared.Rent(newMessageLength);
+                try
+                {
+                    buffer.AsSpan(0, result.Read).CopyTo(newBytes.AsSpan(currentMessageLength, result.Read));
+                    if (!ReferenceEquals(combinedData, Array.Empty<byte>()))
+                    {
+                        combinedData.AsSpan(0, currentMessageLength).CopyTo(newBytes.AsSpan(0, currentMessageLength));
+                    }
+
```

Comment:
Unsure if all those copying is optimal, though most messages should be below 4KB

Reply:
My idea:

1. Get buffer from array pool (4k at start)
2. Read it (4k) from current offset using AsMemory() instead of [], can use that instead of ArraySegment when reading from sockets
3. if message not finished, grow it by some size (4k or probably better some multiplication of 4k TBD)
4. at the end if the buffer is too large, return it to array pool, if its ok in size keep it

Benefits:
a) Reduce copying from buffer->message accumulator (we are reading with AsMemory direcly to message accumulator, they are the same)
b) If we are over 4k, we might be A LOT over 4k, we can grow the buffer with some heuristic (x2 every time like List<> for example), this would reduce times we copy the message over itself

Reply:
I didn't want to change this under this PR because it's already big and I would prefer not to introduce to many changes at the same time. It will be better to deliver smth that works and improve it later.
So would you raise a separet ticket for this one?

Reply:
Created: https://github.com/NethermindEth/nethermind/issues/3278